### PR TITLE
add PipeChannel

### PR DESCRIPTION
**Pull request by weissi**
Motivation:

There are use-cases for SwiftNIO where there are no actual sockets but
rather two pipe file descriptors - one for input, one for output.
There's no real reason why SwiftNIO shouldn't work for those.

Modifications:

Add a PipeChannel.

Result:

More use-cases for SwiftNIO.

### CODE REVIEW

File: Sources/NIO/BaseStreamSocketChannel.swift [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+class BaseStreamSocketChannel<Socket: SocketProtocol>: BaseSocketChannel<Socket> {
```
Comment (weissi) [inline Sources/NIO/BaseStreamSocketChannel.swift]: everything here is moved from `BaseSocketChannel`. Basically the lines that are now left in `BaseSocketChannel` plus `BaseStreamSocketChannel` are pretty much exactly what used to be in `BaseSocketChannel`.

File: Sources/NIO/SocketChannel.swift [hunk 2]
```diff
@@ -32,42 +32,21 @@ private extension ByteBuffer {
 /// A `Channel` for a client socket.
 ///
 /// - note: All operations on `SocketChannel` are thread-safe.
-final class SocketChannel: BaseSocketChannel<Socket> {
-
+final class SocketChannel: BaseStreamSocketChannel<Socket> {
     private var connectTimeout: TimeAmount? = nil
-    private var connectTimeoutScheduled: Scheduled<Void>?
```
Comment (weissi) [inline Sources/NIO/SocketChannel.swift]: all the removed lines here, you'll find in `BaseStreamSocketChannel` above.

File: Tests/NIOTests/StreamChannelsTest.swift [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+            XCTAssertNoThrow(try writeFullyDonePromise.futureResult.wait())
+            // To finish up, let's just tear this down.
+            XCTAssertNoThrow(try chan2.close().wait())
+            XCTAssertNoThrow(try chan1.closeFuture.wait())
+        }
+        XCTAssertNoThrow(try forEachCrossConnectedStreamChannelPair(runTest))
+    }
+
+    ///
+    /// TODO: Re-enable test
```
Comment (weissi) [inline Tests/NIOTests/StreamChannelsTest.swift]: this will need fixing before we can merge.

**Comment by weissi** [comment]
@Lukasa / @normanmaurer this is getting pretty much ready. We need to settle on the correct behaviour for half-close from both sides and I'll add some more tests but the general shape is getting there. Feel free to have an early peek :)

**Comment by weissi** [comment]
At the moment, we're regressing 2 allocations per connection :|.

File: Sources/NIO/BaseSocket.swift [hunk 4]
```diff
@@ -12,6 +12,28 @@
...
 
+import NIOConcurrencyHelpers
+
+private let globallyIgnoredSIGPIPE: Atomic<Bool> = .init(value: false)
+
+func ignoreSIGPIPE(descriptor fd: CInt) throws {
+    if !globallyIgnoredSIGPIPE.load() {
+        #if os(Linux)
+        /* no F_SETNOSIGPIPE on Linux :( */
+        _ = unsafeBitCast(Glibc.signal(SIGPIPE, SIG_IGN) as sighandler_t?, to: Int.self)
```
Comment (Lukasa) [inline Sources/NIO/BaseSocket.swift]: What's the point of an `unsafeBitCast` if we're throwing away the result?

File: Sources/NIO/BaseSocket.swift [hunk 5]
```diff
@@ -12,6 +12,28 @@
...
+        globallyIgnoredSIGPIPE.store(true)
+        #else
+        assert(fd >= 0, "illegal file descriptor \(fd)")
```
Comment (Lukasa) [inline Sources/NIO/BaseSocket.swift]: May as well be a `precondition` tbh.

File: Sources/NIO/BaseSocket.swift [hunk 6]
```diff
@@ -12,6 +12,28 @@
 //
 //===----------------------------------------------------------------------===//
 
+import NIOConcurrencyHelpers
+
+private let globallyIgnoredSIGPIPE: Atomic<Bool> = .init(value: false)
+
+func ignoreSIGPIPE(descriptor fd: CInt) throws {
```
Comment (Lukasa) [inline Sources/NIO/BaseSocket.swift]: Is there any reason for this function not to be namespaced into `BaseSocket`?

File: Sources/NIO/BaseSocket.swift [hunk 7]
```diff
@@ -319,6 +327,7 @@ class BaseSocket: Selectable {
     init(descriptor: CInt) {
         precondition(descriptor >= 0, "invalid file descriptor")
         self.descriptor = descriptor
+        try! ignoreSIGPIPE(descriptor: descriptor)
```
Comment (Lukasa) [inline Sources/NIO/BaseSocket.swift]: Isn't the dance with the atomic above unnecessary? Just do it inside an implicit `dispatch_once` through means of a global or static let.

File: Sources/NIO/BaseStreamSocketChannel.swift (hunk 1 above)
Comment (Lukasa) [reply Sources/NIO/BaseStreamSocketChannel.swift]: That...seems to be untrue? There aren't many deletions in `BaseSocketChannel`. Do you mean everything here is moved from `StreamSocketChannel`?

File: Sources/NIO/Bootstrap.swift [hunk 8]
```diff
@@ -705,3 +705,88 @@ public final class DatagramBootstrap {
...
+    /// - parameters:
+    ///     - option: The option to be applied.
+    ///     - value: The value for the option.
+    @inlinable
+    public func channelOption<Option: ChannelOption>(_ option: Option, value: Option.Value) -> Self {
+        self._channelOptions.append(key: option, value: value)
+        return self
+    }
+
+    /// Create the `PipeChannel` with the provided input and output file descriptors..
```
Comment (Lukasa) [inline Sources/NIO/Bootstrap.swift]: Duplicate period.

File: Sources/NIO/Bootstrap.swift [hunk 9]
```diff
@@ -705,3 +705,88 @@ public final class DatagramBootstrap {
...
+    ///
+    /// - parameters:
+    ///     - inputDescriptor: The _Unix file descriptor_ for the input (ie. the read side).
+    ///     - outputDescriptor: The _Unix file descriptor_ for the input (ie. the write side).
+    /// - returns: an `EventLoopFuture<Channel>` to deliver the `Channel` immediately.
```
Comment (Lukasa) [inline Sources/NIO/Bootstrap.swift]: The word "immediately" here is not accurate: this future may take a while to be delivered if the initializer delays setup.

File: Sources/NIO/PipePair.swift [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+extension NIOFileHandle: Selectable {
+}
```
Comment (Lukasa) [inline Sources/NIO/PipePair.swift]: I assume we don't need this?

File: Sources/NIO/PipePair.swift [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+final class PipePair: SocketProtocol {
+    typealias SocketType = NIOFileHandle
+
+    let inputFD: NIOFileHandle
+    let outputFD: NIOFileHandle
+
+    init(inputFD: NIOFileHandle, outputFD: NIOFileHandle) {
+        self.inputFD = inputFD
+        self.outputFD = outputFD
+        try! inputFD.withUnsafeFileDescriptor {
```
Comment (Lukasa) [inline Sources/NIO/PipePair.swift]: I assume the plan isn't to crash if either of these things fail.

File: Sources/NIO/PipePair.swift [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    func sendmmsg(msgs: UnsafeMutableBufferPointer<MMsgHdr>) throws -> IOResult<Int> {
+        throw ChannelError.operationUnsupported
+    }
+
+    func shutdown(how: Shutdown) throws {
+        switch how {
+        case .RD:
+            throw ChannelError.operationUnsupported
```
Comment (Lukasa) [inline Sources/NIO/PipePair.swift]: Is this a reasonable semantic? Why not close the input FD? Pipes are nearly the only place where `shutdown(.RD)` actually has some remotely observable behaviour.

File: Sources/NIO/PipePair.swift [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+        case .WR:
+            try self.outputFD.close()
+        case .RDWR:
+            try self.close()
+        }
+    }
+
+    var isOpen: Bool {
+        return self.inputFD.isOpen && self.outputFD.isOpen
```
Comment (Lukasa) [inline Sources/NIO/PipePair.swift]: Is this right? Are we not open if _either_ is open? Certainly `close()` thinks we are.

File: Sources/NIO/SocketChannel.swift [hunk 14]
```diff
@@ -12,7 +12,7 @@
 //
 //===----------------------------------------------------------------------===//
 
-private extension ByteBuffer {
+extension ByteBuffer {
     mutating func withMutableWritePointer(body: (UnsafeMutableRawBufferPointer) throws -> IOResult<Int>) rethrows -> IOResult<Int> {
```
Comment (Lukasa) [inline Sources/NIO/SocketChannel.swift]: Add `private` here please, unless you are actually planning to make this more public. If you are, please add it to `BaseStreamSocketChannel.swift`.

File: Sources/NIO/BaseStreamSocketChannel.swift (hunk 1 above)
Comment (weissi) [reply Sources/NIO/BaseStreamSocketChannel.swift]: From `SocketChannel` actually :)

File: Sources/NIO/BaseSocket.swift (hunk 6 above)
Comment (weissi) [reply Sources/NIO/BaseSocket.swift]: done (to `BaseSocketProtocol`)

File: Sources/NIO/BaseSocket.swift (hunk 4 above)
Comment (weissi) [reply Sources/NIO/BaseSocket.swift]: none :)

File: Sources/NIO/BaseSocket.swift (hunk 5 above)
Comment (weissi) [reply Sources/NIO/BaseSocket.swift]: @Lukasa usually we do 'if it's a bug in NIO that doesn't lead to disaster' == `assert`, no? This doesn't cause any harm if `< 0`, `fcntl` will return `EBADF`.

File: Sources/NIO/BaseSocket.swift (hunk 7 above)
Comment (weissi) [reply Sources/NIO/BaseSocket.swift]: no, thanks

File: Sources/NIO/Bootstrap.swift (hunk 8 above)
Comment (weissi) [reply Sources/NIO/Bootstrap.swift]: ta, fixed

File: Sources/NIO/Bootstrap.swift (hunk 9 above)
Comment (weissi) [reply Sources/NIO/Bootstrap.swift]: yeah, sorry, I copy & pasted that from our existing docs, will fix that.

File: Sources/NIO/PipePair.swift (hunk 10 above)
Comment (weissi) [reply Sources/NIO/PipePair.swift]: We do. For `PipePair` to be a `SocketProtocol` implementation, it's `SocketType` must conform to `Selectable`. `PipePair'`s `SocketType` however is `NIOFileHandle`.

File: Sources/NIO/PipePair.swift (hunk 11 above)
Comment (weissi) [reply Sources/NIO/PipePair.swift]: thanks, fixing.

File: Sources/NIO/PipePair.swift (hunk 12 above)
Comment (weissi) [reply Sources/NIO/PipePair.swift]: fair.

File: Sources/NIO/PipePair.swift (hunk 13 above)
Comment (weissi) [reply Sources/NIO/PipePair.swift]: oops :), thanks

File: Sources/NIO/SocketChannel.swift (hunk 14 above)
Comment (weissi) [reply Sources/NIO/SocketChannel.swift]: @Lukasa we're using this from two places now (`BaseStreamSocketChannel.swift` and `SocketChannel.swift`)

File: Sources/NIO/BaseSocket.swift [hunk 15]
```diff
@@ -204,7 +206,8 @@ extension sockaddr_storage {
 /// Base class for sockets.
 ///
 /// This should not be created directly but one of its sub-classes should be used, like `ServerSocket` or `Socket`.
-class BaseSocket: Selectable {
+class BaseSocket: Selectable & BaseSocketProtocol {
```
Comment (Lukasa) [inline Sources/NIO/BaseSocket.swift]: I don't believe in `&` for protocol conformances.

File: Sources/NIO/BaseSocket.swift [hunk 16]
```diff
@@ -319,6 +308,7 @@ class BaseSocket: Selectable {
     init(descriptor: CInt) {
         precondition(descriptor >= 0, "invalid file descriptor")
         self.descriptor = descriptor
+        try! self.ignoreSIGPIPE(descriptor: descriptor)
```
Comment (Lukasa) [inline Sources/NIO/BaseSocket.swift]: Please add a comment explaining why we force-unwrap here.

File: Sources/NIO/PipePair.swift [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+                    atLeastOneSuccess = true
+                }
+            } catch let err as IOError where err.errnoCode == EBADF {
+                // this is what `withUnsafeFileDescriptor` throws if the file handle is closed.
+                error = err
+            } catch {
+                throw error
+            }
+        }
+        guard atLeastOneSuccess else {
```
Comment (Lukasa) [inline Sources/NIO/PipePair.swift]: Why do we tolerate one failure?

File: Sources/NIO/BaseSocket.swift (hunk 5 above)
Comment (weissi) [reply Sources/NIO/BaseSocket.swift]: @Lukasa wdyt?

Comment (Lukasa) [reply Sources/NIO/BaseSocket.swift]: Happy to ignore this then.

File: Sources/NIO/Bootstrap.swift [hunk 18]
```diff
@@ -705,3 +705,114 @@ public final class DatagramBootstrap {
...
+    @inlinable
+    public func channelOption<Option: ChannelOption>(_ option: Option, value: Option.Value) -> Self {
+        self._channelOptions.append(key: option, value: value)
+        return self
+    }
+
+    private func validateFileDescriptorIsNotAFile(_ descriptor: CInt) throws {
+        var s: stat = .init()
+        try withUnsafeMutablePointer(to: &s) { ptr in
+            try Posix.fstat(descriptor: descriptor, outStat: ptr)
```
Comment (weissi) [inline Sources/NIO/Bootstrap.swift]: @Lukasa this is currently run on the EventLoop. I think I might actually do what you suggested and for now just assert that it's not run on the EventLoop. We can always improve later by allowing that by offloading it.

File: Sources/NIO/Bootstrap.swift [hunk 19]
```diff
@@ -705,3 +705,114 @@ public final class DatagramBootstrap {
...
+        }
+        if (s.st_mode & S_IFREG) != 0 {
```
Comment (weissi) [inline Sources/NIO/Bootstrap.swift]: this should also check for `S_IFDIR`

**Comment by weissi** [comment]
ping @Lukasa

File: Sources/NIO/Bootstrap.swift [hunk 20]
```diff
@@ -705,3 +705,116 @@ public final class DatagramBootstrap {
...
+///                       }
+///                       .withPipes(inputDescriptor: STDIN_FILENO, outputDescriptor: STDOUT_FILENO)
+///
+public final class NIOPipeBootstrap {
+    private let group: EventLoopGroup
+    private var channelInitializer: ((Channel) -> EventLoopFuture<Void>)?
+    @usableFromInline
+    internal var _channelOptions = ChannelOptions.Storage()
+
+    /// Create a `PipeBootstrap` on the `EventLoopGroup` `group`.
```
Comment (glbrntt) [inline Sources/NIO/Bootstrap.swift]:
```suggestion
    /// Create a `NIOPipeBootstrap` on the `EventLoopGroup` `group`.
```

Comment (glbrntt) [reply Sources/NIO/Bootstrap.swift]: And elsewhere.

File: Sources/NIO/Bootstrap.swift [hunk 21]
```diff
@@ -705,3 +705,116 @@ public final class DatagramBootstrap {
...
+
+    /// Initialize the connected `PipeChannel` with `initializer`. The most common task in initializer is to add
+    /// `ChannelHandler`s to the `ChannelPipeline`.
+    ///
+    /// The connected `Channel` will operate on `ByteBuffer` as inbound and outbound messages. Please note that
+    /// `IOData.fileRegion` is _not_ supported for `PipeChannel`s because `sendfile` only works on sockets.
+    ///
+    /// - parameters:
+    ///     - handler: A closure that initializes the provided `Channel`.
+    public func channelInitializer(_ handler: @escaping (Channel) -> EventLoopFuture<Void>) -> Self {
```
Comment (glbrntt) [inline Sources/NIO/Bootstrap.swift]: Isn't `Self` Swift 5.1?

File: Sources/NIO/Bootstrap.swift [hunk 22]
```diff
@@ -705,3 +705,116 @@ public final class DatagramBootstrap {
...
+        if (s.st_mode & S_IFREG) != 0 || (s.st_mode & S_IFDIR) != 0 {
+            throw ChannelError.operationUnsupported
+        }
+    }
+
+    /// Create the `PipeChannel` with the provided input and output file descriptors.
+    ///
+    /// - parameters:
+    ///     - inputDescriptor: The _Unix file descriptor_ for the input (ie. the read side).
+    ///     - outputDescriptor: The _Unix file descriptor_ for the input (ie. the write side).
```
Comment (glbrntt) [inline Sources/NIO/Bootstrap.swift]:
```suggestion
    ///     - outputDescriptor: The _Unix file descriptor_ for the output (ie. the write side).
```

File: Sources/NIO/PipePair.swift [hunk 23]
```diff
@@ -0,0 +1,137 @@
...
+            }
+        }
+    }
+
+    var description: String {
+        return "PipePair { in=\(inputFD), out=\(outputFD) }"
+    }
+
+    func connect(to address: SocketAddress) throws -> Bool {
+        return true
```
Comment (glbrntt) [inline Sources/NIO/PipePair.swift]: Why isn't this a `ChannelError.operationUnsupported`?

Comment (weissi) [reply Sources/NIO/PipePair.swift]: thanks, no reason I think, fixing.

File: Sources/NIO/Bootstrap.swift (hunk 21 above)
Comment (weissi) [reply Sources/NIO/Bootstrap.swift]: for return types, it was there for a long time

File: Sources/NIO/Bootstrap.swift (hunk 22 above)
Comment (weissi) [reply Sources/NIO/Bootstrap.swift]: thansk!

**Comment by weissi** [comment]
@swift-nio-bot test perf please

**Pull request merged**